---
title: 'Using jackstraw to evaluate cluster membership in single cell data'
author: "Neo Christopher Chung"
output:
  knitr:::html_vignette:
    toc: true
vignette: |
  %\VignetteIndexEntry{tms: Temporal Molecular Signatures}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r global options, include = FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, cache=TRUE, fig.height = 4, fig.width = 7, fig.align = "center",tidy.opts=list(width.cutoff=80),tidy=TRUE)
```

This vignette is largely derived from the publication in Bioinformatics. For more details, please read "Statistical significance of cluster membership for unsupervised evaluation of cell identities" at https://doi.org/10.1093/bioinformatics/btaa087

We demonstrate **how the jackstraw test for cluster membership is applied on a mixture of Jurkat and 293T cell lines.**

## Abstract

Single-cell RNA-sequencing (scRNA-seq) allows us to dissect transcriptional heterogeneity arising from cellular types, spatio-temporal contexts and environmental stimuli. Transcriptional heterogeneity may reflect phenotypes and molecular signatures that are often unmeasured or unknown a priori. Cell identities of samples derived from heterogeneous subpopulations are then determined by clustering of scRNA-seq data. These cell identities are used in downstream analyses. How can we examine if cell identities are accurately inferred? Unlike external measurements or labels for single cells, using clustering-based cell identities result in spurious signals and false discoveries.

We introduce non-parametric methods to evaluate cell identities by testing cluster memberships in an unsupervised manner. Diverse simulation studies demonstrate accuracy of the jackstraw test for cluster membership. We propose a posterior probability that a cell should be included in that clustering-based subpopulation. Posterior inclusion probabilities (PIPs) for cluster memberships can be used to select and visualize samples relevant to subpopulations.

![Visual explanation for circular analysis in naively evaluating cluster memberships. In this example, scRNA-seq data are clustered to obtain K=2 cellular subpopulations. Since cell identities are estimated by clustering scRNA-seq data, testing if a cell is correctly assigned to its presumed cellular subpopulation results in artificially inflated significance. The proposed jackstraw for clustering overcomes this challenge by learning the overfitting inherent in evaluating cluster membership](jackstraw-cluster-fig/CircularAnalysis.png){#id .class width=400 height=240}

## Applying the jackstraw test for cluster membership to Jurkat and 293T cell lines

Cells from a mixture of Jurkat and 293T cell lines (50:50) were sequenced using GemCode by 10Ã— Genomics (Zheng et al., 2017). Jurkat and 293T cell lines are highly distinct, being derived from male and female individuals, respectively. Zheng et al. (2017) applied K-means clustering that separates m=3381 cells into K=2 subpopulations. Following quality control, normalization, gene selection and dimension reduction in the original analysis, we tested whether individual cells are correctly assigned to one of two subpopulations based on the top 10 PCs of UMI.

Load the first 10 PCs of 293T:Jurkat data from Supplementary Data 1 of Zheng et al., 2017. Then we apply and visualize K-means clustering:

```{r clustering}
library(jackstraw)
library(ggplot2)

dat <- read.csv("ncomms14049-s2.csv")
dat <- dat[dat$Sample == "50%:50% Jurkat:293T",]
dat <- dat[,-1]
Jurkat293T <- dat

set.seed(0)
dat.plot <- dat
k <- kmeans(dat,2,iter.max=100, nstart = 100)
dat.plot$k <- k$cluster
g <- ggplot(dat.plot,aes(PC1,PC2,col=as.factor(k)))+geom_point(size=0.2,alpha=1)+theme_classic()+
   scale_color_manual(values=c("navyblue","darkorange"))
print(g)
```

We noticed that while the most of cells seem to be well contained in 2 separated clusters, some cell identities appear to be more ambiguous. We are interested in evaluating those "in-between" cells.

Apply the jackstraw test for cluster membership to obtain p-values. The important parameters are ```s``` and ```B```. In this example, we used s=.1m and B=1000, where ```m``` is the number of cells. Generally, keep ```s``` small compared to ```m```. At the same value of ```s```, a bigger ```B``` leads to a higher resolution of p-values. For the further discussion on this, please see ```2.1 Jackstraw test for cluster membership``` on Chung (2020).

```{r jackstraw_cluster}
set.seed(0) 
js <- jackstraw_kmeans(as.matrix(dat), kmeans.dat=k, s=round(nrow(dat)*.1), B=1000, verbose=FALSE, pool=FALSE)

dat.plot <- dat
dat.plot$k <- k$cluster
dat.plot$Pvalue <- js$p.F

ggplot(dat.plot,aes(Pvalue))+
  geom_histogram(bins = 30) +
  theme_minimal() 

g.pvalue <- ggplot(dat.plot,aes(PC1,Pvalue))+
  geom_point(aes(col=as.factor(k),shape=as.factor(k)), size=0.6) +
  scale_colour_manual(guide = guide_legend(title = "Cluster",order = 1), labels = c("1","2"), values = c("navyblue","darkorange"))+
  scale_shape_manual(guide = guide_legend(title = "Cluster",order = 1), labels = c("1","2"), values = c(16,17)) +
  theme_minimal() + theme(legend.position="none")
print(g.pvalue)
```

## Estimating ```\pi_0``` and posterior inclusion probabilities (PIPs)

We are interested in estimating ```\pi_0```, which is the number of null cells. Quantitatively, there are a few different ways to estimate. We recommend following ```qvalue``` R/Biocondutor package. However, a hyperparameter (lambda) must be estimated and sometimes it could be problematic. e.g., see the discussion on https://github.com/StoreyLab/qvalue/issues/19

For demonstration purposes, we apply different methods from packages ```mutoss``` and ```qvalue```. Note when lambda is not given to ```qvalue::pi0est```, it's automatically esitmated from fitting a cubic smoothing spline.

```{r est_pi0}
library(qvalue)
# lambda estimated from fitting a cubic smoothing spline
qvalue::pi0est(js$p.F)$pi0

# alternatively, use other estimators such as the Blanchard-Roquain procedure.
library(mutoss)
mutoss::BR_pi0_est(js$p.F, alpha=.3)$pi0

pi0_avg = (qvalue::pi0est(js$p.F)$pi0 + mutoss::BR_pi0_est(js$p.F, alpha=.3)$pi0)/2
```

Here, we use the mean of two results. With the estimated ```pi0_avg```, we run the PIP calculation. And visuaslize p-values and PIPs with ggplot2:

```{r visualize}
pipout = pip(js$p.F, group=k$cluster, pi0=pi0_avg)

library(Matrix)
require(gridExtra)
library(cowplot) 

dat.plot <- dat
dat.plot$k <- k$cluster
dat.plot$PIP <- pipout
dat.plot$Pvalue <- js$p.F

g.pip <- ggplot(dat.plot,aes(PC1,PIP))+
  geom_point(aes(col=as.factor(k),shape=as.factor(k)), size=0.6) +
  scale_colour_manual(guide = guide_legend(title = "Cluster",order = 1), labels = c("1","2"), values = c("navyblue","darkorange"))+
  scale_shape_manual(guide = guide_legend(title = "Cluster",order = 1), labels = c("1","2"), values = c(16,17)) +
  theme_minimal() + theme(legend.position="none")
g.pvalue.pip <- ggplot(dat.plot,aes(Pvalue,PIP))+
  geom_point(aes(col=as.factor(k),shape=as.factor(k)), size=0.6) +
  geom_hline(yintercept = .8, col="red", lty="dashed") +
  scale_x_continuous(limits = c(0,1)) +
  scale_colour_manual(guide = guide_legend(title = "Cluster",order = 1), labels = c("1","2"), values = c("navyblue","darkorange"))+
  scale_shape_manual(guide = guide_legend(title = "Cluster",order = 1), labels = c("1","2"), values = c(16,17)) +
  theme_minimal() + theme(legend.position="none")
print(g.pvalue.pip)

require(scales)
highcol <- "black"
lowcol <- "white"

g.pca <- ggplot(dat.plot,aes(PC1,PC2,alpha=PIP))+
  geom_point(aes(alpha=PIP,col=as.factor(k),shape=as.factor(k)), size=0.6) +
  scale_colour_manual(guide = guide_legend(title = "Cluster",order = 1, override.aes = list(size=1.5)), labels = c("1","2"), values = c("navyblue","darkorange"))+
  scale_shape_manual(guide = guide_legend(title = "Cluster",order = 1), labels = c("1","2"), values = c(16,17)) +
  geom_point(aes(fill = PIP), alpha = 0) +
  scale_fill_gradient(high = highcol, low = lowcol) +
  guides(alpha = F) +
  theme_minimal() + theme(legend.position = "right")
print(g.pca)
```


## References

Chung (2020) Statistical significance of cluster membership for unsupervised evaluation of cell identities. Bioinformatics. https://doi.org/10.1093/bioinformatics/btaa087

Zheng G.X. et al.  (2017) Massively parallel digital transcriptional profiling of single cells. Nat. Commun.  https://www.nature.com/articles/ncomms14049
